### PR TITLE
feat(cli/tools/upgrade): canary support

### PR DESCRIPTION
**Pull request by crowlKats**

### CODE REVIEW

File: cli/version.rs [hunk 1]
```diff
@@ -1,9 +1,15 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
 
-pub const DENO: &str = env!("CARGO_PKG_VERSION");
 pub const GIT_COMMIT_HASH: &str = env!("GIT_COMMIT_HASH");
 pub const TYPESCRIPT: &str = crate::js::TS_VERSION;
 
+pub fn deno() -> String {
+  let semver = env!("CARGO_PKG_VERSION");
+  option_env!("DENO_CANARY").map_or(semver.to_string(), |_| {
```
Comment (bartlomieju) [inline cli/version.rs:8]: Is `DENO_CANARY` env variable defined anywhere?

Comment (crowlKats) [reply cli/version.rs:8]: as discussed on the discord with luca, it will be added to ci

File: cli/flags.rs [hunk 2]
```diff
@@ -951,6 +954,11 @@ update to a different location, use the --output flag
         .short("f")
         .help("Replace current exe even if not out-of-date"),
     )
+    .arg(
+      Arg::with_name("canary")
+        .long("canary")
+        .help("Upgrade to canary builds"),
```
Comment (bartlomieju) [inline cli/flags.rs:960]: IMO this arg should take optional value that allows to specify concrete commit

Comment (crowlKats) [reply cli/flags.rs:960]: with the current state of the PR, that would be passed through `version`

**Comment by lucacasonato** [comment]
@crowlKats Can you add a test?

**Comment by crowlKats** [comment]
@lucacasonato sure; but currently all upgrade tests seem to be disabled because flaky

**Comment by bartlomieju** [comment]
> @lucacasonato sure; but currently all upgrade tests seem to be disabled because flaky

@crowlKats could you uncomment them and try to make them pass?

**Comment by crowlKats** [comment]
not really sure what's going on... it isnt doing anything special...

**Comment by crowlKats** [comment]
Description of current behaviour of upgrade with this PR:
to specify hash for canary, set `version` flag.

`version` specified:
stable or canary to canary: passed hash must not be same as current hash
stable to stable: same as currently (version must not be older or same as current)
canary to stable: will pass no matter what (no checks)

`version` not specified:
stable or canary to canary: latest hash must not be same as current hash
stable to stable: same as currently (latest version must not be older or same as current)
canary to stable: will pass no matter what (no checks)

these rules of course can be bypassed by `force`.

**Comment by lucacasonato** [comment]
The behavior looks good to me. It seems the CI failure outs m is unrelated.

**Review by lucacasonato — APPROVED** [review]
LGTM - thanks @crowlKats!

**Pull request merged**